File: 03 Making Regression Make Sense/Figure 3-1-2.py
#!/usr/bin/env python
"""
Tested on Python 3.4
"""

import urllib
import zipfile
import urllib.request
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt

# Download data and unzip the data
urllib.request.urlretrieve('http://economics.mit.edu/files/397', 'asciiqob.zip')
with zipfile.ZipFile('asciiqob.zip', "r") as z:
   z.extractall()

# Read the data into a pandas dataframe
pums         = pd.read_csv('asciiqob.txt',
	                       header           = None,
	                       delim_whitespace = True)
pums.columns = ['lwklywge', 'educ', 'yob', 'qob', 'pob']

# Set up the model
y = pums.lwklywge
X = pums.educ
X = sm.add_constant(X)

# Save coefficient on education
model      = sm.OLS(y, X)
results    = model.fit()
educ_coef  = results.params[1]
intercept  = results.params[0]

# Calculate means by educ attainment and predicted values
groupbyeduc = pums.groupby('educ')
educ_means  = groupbyeduc['lwklywge'].mean()
yhat        = pd.Series(intercept + educ_coef * educ_means.index.values, 
                        index = educ_means.index.values)

# Create plot
plt.figure()
educ_means.plot('-o')
yhat.plot()
plt.xlabel("Years of completed education")
plt.ylabel("Log weekly earnings, \$2003")
plt.legend().set_visible(False)
plt.savefig('Figure 3-1-2-Python.pdf')

# End of script

File: 03 Making Regression Make Sense/Figure 3-1-2.r
# R code for Figure 3.1.2               #
# Required packages                     #
# - ggplot2: making pretty graphs       #
# - data.table: simple way to aggregate #
library(ggplot2)
library(data.table)

# Download data and unzip the data
download.file('http://economics.mit.edu/files/397', 'asciiqob.zip')
unzip('asciiqob.zip')

# Read the data into a dataframe
pums        <- read.table('asciiqob.txt',
                          header           = FALSE,
                          stringsAsFactors = FALSE)
names(pums) <- c('lwklywge', 'educ', 'yob', 'qob', 'pob')

# Estimate OLS regression
reg.model <- lm(lwklywge ~ educ, data = pums)

# Calculate means by educ attainment and predicted values
pums.data.table <- data.table(pums)
educ.means      <- pums.data.table[ , list(mean = mean(lwklywge)), by = educ]
educ.means$yhat <- predict(reg.model, educ.means)

# Create plot
p <- ggplot(data = educ.means, aes(x = educ)) +
     geom_point(aes(y = mean))                +
     geom_line(aes(y = mean))                 +
     geom_line(aes(y = yhat))                 +
     ylab("Log weekly earnings, $2003")       +
     xlab("Years of completed education")

ggsave(filename = "Figure 3-1-2-R.pdf")


# End of file
